### DISCUSSION TITLE
Complicated structure, nested contents with interface type and derived fields

### DISCUSSION THREAD

**Discussion by venkatakrishnan-tu** [graphql-java/graphql-java#2756] [Q&A]
Currently utilizing PropertyDataFetcher to populate fields in nested schema. Have the need to introduce polymorphism to find the actual type of object at run time and fetch the data. As the root level is fetched through property, there is no way now to introduce data fetcher for the nested interface.  I morphed the object structure. 

type UniversityAlumni{
id
name
positions: [Position]
qualifications: [Qualification]
}

type Qualification{

id
name
activeSince                          //calculated from the veryFirstDegree
awards                                 //calculated from all the degrees for this alumni
duration
totalNumberOfcredits                 //calculated from all the degrees for this alumni
lastGraduatedDate                    //calculated from the mostRecentDegree

veryFirstDegree: [Degree]
mostRecentDegree: [Degree]
}

interface Degree{
    name
    abbreviation  
    credits
}

type bachelorDegree implements Degree{
    name
    abbreviation  
    department  
}

type mastersDegree implements Degree{
    name
    abbreviation  
    speciality
    thesisEligible
}

Currently, I have single datafetcher for UniversityAlumni, that maps the fields using implicit PropertyDataFetcher
Due to the level complexity in forming Qualification, couldn't add data fetcher for each field. **Because, the child <Degree> has to be resolved first in order to populate the derived field in Qualification....** 

Question 1: Is there any other way to acheive this?? 

--------------------
Question 2: How to achieve the below query? I can't wire datafetcher for interface without the qualification data.  ? The derived field in qualifications is making it complicated to wire the fetcher for interface.

query UniversityAlumni{
id
name
qualifications {
    name 
    activeSince           
    awards                
    duration
    totalNumberOfcredits  
    lastGraduatedDate     

    veryFirstDegree {
      name
      abbreviation  
      credits

      ... on BachelorDegree {
          department
      }

       ... on MastersDegree {
          speciality
      }

    }

    mostRecentDegree{
      name
      abbreviation  
      credits

      ... on BachelorDegree {
          abbreviation
          department
      }

       ... on MastersDegree {
          thesisEligible
      }

    }

}

Please help.

**Comment by jord1e** [comment]
You are probably looking for selection sets; https://www.graphql-java.com/documentation/field-selection

the datafetcher would be on Qualification and you’d go from there

**Reply by venkatakrishnan-tu** [reply]
Thanks @jord1e for quick response. 

Are you saying, do data fetcher on qualification, then do data mapping? 
https://www.graphql-java.com/documentation/data-mapping/


In the above structure, Qualification is two level inside from the top, <UniversityAlumni>, the UniversityAlumni will also need to be queried. Per my current implementation, datafetcher is at UniversityAlumni level.  If I were to change the data fetcher to Qualification level,  I want to understand which is the trigger point.. 

I would need list of degrees, in order to form the derived fields. And since they are interface,  the type resolver need to be registered upfront. How can I achieve that?

**Reply by venkatakrishnan-tu** [reply]
Also, one more thing to note.. lets say somehow this Qualification is obtained/data mapped.. the typeresolver and its data fetcher on Degree<interface> seems redundant.? 
cc:  @andimarek @bbakerman
